### DISCUSSION TITLE
Add Support for Obsidian Callouts (Admonitions)

### DISCUSSION THREAD

**Discussion by RoyiAvital** [srid/emanote#307] [Ideas]
Lately [Obsidian has added support for Callouts (Admonitions)](https://help.obsidian.md/How+to/Use+callouts).

Though the didn't chose the most used syntax, it would be great to add support for this.

I'd go even farther and strive for feature / syntax compatibility with Obsidian.

**Comment by toastal** [comment]
Obsidian's syntax is mostly compatible with [Microsoft's Markdown](https://learn.microsoft.com/en-us/contribute/markdown-reference#alerts-note-tip-important-caution-warning) but will probably conflict with admonitions in Markdown for [CommonMark RFC](https://talk.commonmark.org/t/generic-directives-plugins-syntax/444)+[Docusaurus](https://docusaurus.io/docs/markdown-features/admonitions)+[Remark](https://github.com/remarkjs/remark-directive), [MkDocs](https://squidfunk.github.io/mkdocs-material/reference/admonitions/), [Mozilla Developer Network](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Howto/Markdown_in_MDN#notes_warnings_and_callouts), and [Microsoft GitHub's broken implimentation](https://github.com/orgs/community/discussions/16925).

But at least admonitions are standardized in [AsciiDoc](https://docs.asciidoctor.org/asciidoc/latest/blocks/admonitions/) and [reStructuredText](https://docutils.sourceforge.io/docs/ref/rst/directives.html#admonitions).

**Comment by RoyiAvital** [comment]
Maybe a YAML to configure the syntax for them?
Anyhow, I think Obsidian compatibility, at least under a flag, is a major feature.

**Reply by toastal** [reply]
And I would say AsciiDoc support should be a major feature 😅. It could be less trouble that trying to support 5-6 different Markdown syntax forks.

**Reply by RoyiAvital** [reply]
The idea I raised is that `emanote` main feature would be a static site generator for personal knowledge managers.
So you need to chose one, better one which is popular, and have some kind of a compatibility with it.

If Obsidian supported AsciiDoc I would be the first to support.
But none of the PKM's I'm aware of supports AsciiDoc.
Do you know any?

**Reply by srid** [reply]
[If](https://x.com/sridca/status/1891504629126414818) there was a complete AsciiDoc Haskell parser library, I'd readily add AsciiDoc support to Emanote.

**Comment by srid** [comment]
It is time to implement this: https://github.com/srid/emanote/issues/465

And I think we should choose one format and stick with it. I propose to go with Obsidian.

**Reply by toastal** [reply]
I would suggest the [CommonMark RFC](https://talk.commonmark.org/t/generic-directives-plugins-syntax/444)+[Docusaurus](https://docusaurus.io/docs/markdown-features/admonitions)+[Remark](https://github.com/remarkjs/remark-directive) route as it doesn’t overload the blockquote sigil `>` (breaks when rendered in another tool, & [goes against CommonMark](https://spec.commonmark.org/0.30/#block-quotes)), follows the blocks RFC, aren’t a part of a proprietary Markdown fork, & Pandoc already understands `:::` blocks. The biggest downside is that the keywords are English which makes writing in other languages jarring (but this is a problem with most implementations, including Obsidian).

**Reply by RoyiAvital** [reply]
I agree with @toastal . It must work as close can be to standards.  
The option to support the Obsidian syntax should be an Opt In option.

Though, in my case, this is what I will be using as I basically use `emanote` as a site generator for Obsidian.

**Reply by srid** [reply]
I'll consider supporting _both_. 

(We are migrating a Docusaurus site to Emanote, so this would be useful actually)

**Reply by srid** [reply]
One problem with docusaurus format is we can't support callouts with custom types: https://docusaurus.io/docs/markdown-features/admonitions#specifying-title

Because `commonmark-hs` (the Markdown parser used by Emanote) doesn't recognize that syntax. ie., `Pandoc already understands ::: blocks` is not true if those blocks contain a custom title.

**Reply by srid** [reply]
Demo: https://nixos.asia/en/blog/nix-rapid

Draft PR: https://github.com/srid/emanote/pull/466

**Reply by toastal** [reply]
Consider adding `role="note"` to callouts for better semantics/accessibility

**Comment by nramirezuy** [comment]
Do this feature has to be limited to a [pre-configured set](https://github.com/srid/emanote/blob/master/emanote/src/Emanote/Pandoc/Renderer/Callout.hs#L49) on callouts?

I just started using **emanote** and I already found myself requiring to add _Quotes_. I was pretty disappointing figuring out that those weren't added yet .

So I went ahead and created [a JS module](https://gist.github.com/nramirezuy/3f669efab672530d27d7256497a26d97) to add them myself. While doing so, I can't but to ask myself: _why it wasn't implemented generic?_

In my module implementation I added a default icon for the notation, and always render a callout if I see the notation. But it could also fallback to standard HTML _blockquote_.

![image](https://github.com/user-attachments/assets/6f8e8a6b-5a0c-4082-a12d-a5762b8b1a50)

**Reply by srid** [reply]
@nramirezuy What is the associated Markdown for those screenshots? And how does Obsidian render them?

**Reply by nramirezuy** [reply]
Here's the markdown associated:

```markdown
> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

> [!default]
>
> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

> [!default] Default with Title
>
> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

> [!quote]
>
> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.

> [!quote] Quote with Title
>
> Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
```

I stole _Quote_ from [Obsidian](https://help.obsidian.md/Editing+and+formatting/Callouts):
![image](https://github.com/user-attachments/assets/0a440e9b-fb3d-401b-8655-7742e20260e8)

_Default_ is my own invention:

![image](https://github.com/user-attachments/assets/553d87a4-9bb5-47f5-bc9a-4e58a277c3fb)

Now looking at the [Obsidian documentation](https://help.obsidian.md/Editing+and+formatting/Callouts#Customize+callouts), I got an idea of changing [callout.tpl](https://github.com/srid/emanote/blob/master/emanote/default/templates/filters/callout.tpl) to use a css value instead. This would mean porting what is already there in **emanote**.

![image](https://github.com/user-attachments/assets/5bdaa12c-1e3b-47d1-9459-c57a5c0aa818)

**Reply by srid** [reply]
It certainly would be good to support custom callouts. PR on that is certainly welcome.

Incidentally, I recently had a need for `quote` callout, so I went ahead and added it: https://github.com/srid/emanote/pull/571

**Reply by srid** [reply]
@nramirezuy You can now add custom callouts easily. Try it out: https://github.com/srid/emanote/pull/573

**Reply by srid** [reply]
Example: https://github.com/srid/srid/commit/14b6e48e1743e9e807573f0ca194f5d41ae7f454 (which is rendered https://srid.ca/harmless )

**Reply by nramirezuy** [reply]
@srid looking good! thanks!
